hazelcast/src/main/java/com/hazelcast/nio/tcp/AbstractSelectionHandler.java

```diff
@@ -46,15 +47,43 @@ public AbstractSelectionHandler(TcpIpConnection connection, IOSelector ioSelecto
         this.initialOps = initialOps;
     }
 
+    public void migrate(final IOSelector newIOSelector) {
+        if (ioSelector == newIOSelector) {
+            return;
+        }
+
+        ioSelector.addTask(new Runnable() {
+            @Override
+            public void run() {
+                unregisterOp(initialOps);
+                ioSelector = newIOSelector;
+                newIOSelector.addTask(new Runnable() {
+                    @Override
+                    public void run() {
+                        selector = newIOSelector.getSelector();
+                        selectionKey = registerSelectionKey();
+                        registerOp(initialOps);
+                    }
+                });
+                newIOSelector.wakeup();
+            }
+        });
+        ioSelector.wakeup();
+    }
+
     protected SelectionKey getSelectionKey() {
         if (selectionKey == null) {
-            try {
-                selectionKey = socketChannel.register(selector, initialOps, this);
-            } catch (ClosedChannelException e) {
-                handleSocketException(e);
-            }
+            selectionKey = registerSelectionKey();
         }
+        return selectionKey;
+    }
 
+    protected SelectionKey registerSelectionKey() {
```

Comment:
We have 2 methods to get a selection key; registerSelectionKey and getSelectionKey. I'm sure they are useful in a certain context; can you add some documentation when they should be used because now it is a bit of magic.